### ISSUE TITLE
Array~unshift was buggy in some old environments, was not generic

### ISSUE THREAD

**Issue by Xotic750** [es-shims/es5-shim#339]
Tests
```
affirm.doesNotThrow(function () {
  [].unshift(true);
});

var unshiftArr = [],
  unshiftObj = {},
  arrCmp,
  i = 0;

affirm.strictEqual(unshiftArr.unshift(undefined), 1, 'array wrong return count');
affirm.strictEqual(unshiftArr.length, 1, 'array length incorrect');
affirm.ok(Object.prototype.hasOwnProperty.call(unshiftArr, 0), 'array value not set');
affirm.strictEqual(unshiftArr[0], undefined, 'array value incorrect');

affirm.strictEqual(Array.prototype.unshift.call(unshiftObj, undefined), 1, 'object wrong return count');
affirm.strictEqual(unshiftObj.length, 1, 'object length incorrect');
affirm.ok(Object.prototype.hasOwnProperty.call(unshiftObj, 0), 'object value not set');
affirm.strictEqual(unshiftObj[0], undefined, 'object value incorrect');

affirm.strictEqual(unshiftArr.unshift(0), 2, 'array wrong return count');
affirm.strictEqual(unshiftArr.length, 2, 'array length incorrect');
affirm.ok(Object.prototype.hasOwnProperty.call(unshiftArr, 0), 'array value not set');
affirm.strictEqual(unshiftArr[0], 0, 'array value incorrect');

affirm.strictEqual(Array.prototype.unshift.call(unshiftObj, 0), 2, 'object wrong return count');
affirm.strictEqual(unshiftObj.length, 2, 'object length incorrect');
affirm.ok(Object.prototype.hasOwnProperty.call(unshiftObj, 0), 'object value not set');
affirm.strictEqual(unshiftObj[0], 0, 'object value incorrect');

arrCmp = [];
arrCmp.length = 12;
arrCmp[0] = undefined;
arrCmp[1] = null;
arrCmp[2] = -1;
arrCmp[3] = 0;
arrCmp[4] = 1;
arrCmp[5] = false;
arrCmp[6] = true;
arrCmp[7] = undefined;
arrCmp[8] = '';
arrCmp[9] = 'abc';
arrCmp[10] = null;
arrCmp[11] = undefined;

unshiftArr = [];
affirm.strictEqual(unshiftArr.unshift(undefined),  1, 'test1');
affirm.strictEqual(unshiftArr.unshift(null),  2, 'test2');
affirm.strictEqual(unshiftArr.unshift('abc'), 3, 'test3');
affirm.strictEqual(unshiftArr.unshift(''), 4, 'test4');
affirm.strictEqual(unshiftArr.unshift(undefined), 5, 'test5');
affirm.strictEqual(unshiftArr.unshift(5), 'test6');
affirm.strictEqual(unshiftArr.unshift(5), 'test7');
affirm.strictEqual(unshiftArr.unshift(5), 'test8');
affirm.strictEqual(unshiftArr.unshift(5), 'test9');
affirm.strictEqual(unshiftArr.unshift(true), 6, 'test10');
affirm.strictEqual(unshiftArr.unshift(false),  7, 'test11');
affirm.strictEqual(unshiftArr.unshift(1), 8, 'test12');
affirm.strictEqual(unshiftArr.unshift(0), 9, 'test13');
affirm.strictEqual(unshiftArr.unshift(-1), 10, 'test14');
affirm.strictEqual(unshiftArr.unshift(null), 11, 'test15');
affirm.strictEqual(unshiftArr.unshift(undefined), 12, 'test16');
affirm.strictEqual(unshiftArr.length, arrCmp.length, 'test17');
while (i < unshiftArr.length) {
  affirm.ok(Object.prototype.hasOwnProperty.call(unshiftArr, i), 'hasOwn test' + i);
  affirm.strictEqual(unshiftArr[i], arrCmp[i], 'strictEqual test' + i);
  i += 1;
}
```
Fix
```
Array.prototype.unshift = function () {
  var object = Object(this),
    len = object.length >>> 0,
    k = len,
    argCount = arguments.length,
    from,
    to,
    j = 0;
  object.length = len + argCount;
  while (k > 0) {
    from = k - 1;
    to = k + argCount - 1;
    if (Object.prototype.hasOwnProperty.call(object, from)) {
      object[to] = object[from];
    } else {
      delete object[to};
    }
    k -= 1;
  }
  while (j < argCount) {
    object[j] = arguments[j];
    j += 1
  }
  return object.length;
};
```
http://www.ecma-international.org/ecma-262/5.1/#sec-15.4.4.13

**Comment by Xotic750** [comment]
A report mentioning IE < 8 problem with `unshift`
https://www.nczonline.net/blog/2010/02/02/how-internet-explorer-8-document-mode-affects-javascript/

**Comment by ljharb** [comment]
This was fixed in 78dd9557 in 2012.
